### DISCUSSION TITLE
Speech final sometimes never returns True

### DISCUSSION THREAD

**Discussion by anunayajoshi** [deepgram/community#409] [General help]
I have been using both interim results and speech final for my streaming pipeline. It works well but sometimes speech final isn't set to true, which causes my pipeline to end up waiting for a finalized transcription which doesn't come , until the user utters something else and it captures that as final, and then sends the entire buffer. 

Is there a way around that? I was imagining adding another time based endpointing. If my deepgram endpointing is set to 200ms, and I don't receive a speech final within 300ms, I send whatever I have in my interim results buffer.

**Comment by jjmaldonis** [comment]
Hey @anunayajoshi, check out the discussion in this post: https://github.com/orgs/deepgram/discussions/384 The solution using `utterance_end_ms` may work well for your use case. You could set it to 200 or 300 ms and it would effectively do exactly what you described.

I have a followup question as well: If the user is not speaking, do you continue to send audio to Deepgram? Or do you know when the speaker stops speaking and pause the message-sending?

**Reply by anunayajoshi** [reply]
Thanks for the quick response! 

Just to confirm, since I can't find any documentation about this parameter `utterance_end_ms` anywhere other than that thread, the parameter only supports 1000,2000,3000ms etc, is that right?

And yes, if the user is not speaking, I am sending audio to Deepgram. My current setup is a telephony server, so sending the audio received from Twilio to deepgram directly. I just have another class checking the results from deepgram, and outputting the final transcript to process.

**Reply by jjmaldonis** [reply]
Thanks for the additional info, the `utterance_end_ms` is likely the best option for your use case.

The `utterance_end_ms` param can be set to any value, and we recommend using a value >= 1000 ms.

**Comment by nikolawhallon** [comment]
Deepgram should be sending final results (`is_final=true`) once every 3-5 seconds, and `speech_final` results only after someone finishes speaking. By default, with `endpointing=true`, it will determine if someone finishes speaking if there is 10ms of silence after they finish speaking - very short, but also very low latency responses. If you set `endpointing=200` and someone finishes speaking, if there is silence after that for 200ms, a `speech_final` result will be sent, but if there is any blip in the audio during that time (someone opens a door, there is a microphone click, etc) then the endpointing algorithm will fail - still, you should see an `is_final` message within 3-5 seconds (if you do not see this, let me know, it could be another issue).

To get around this limitation with endpointing failing due to sounds in the background above the audio's noise profile, we have an algorithm you can turn on with `utterance_end_ms=1000&interim_results=true` which looks for gaps of `utterance_end_ms` duration between words to determine when someone has finished speaking - it only has a resolution of 1s (so values of 1000ms, 2000ms, 3000ms are reasonable) because it relies on inferencing the audio to look for spoken words (via interim results).

**Reply by cateas** [reply]
@nikolawhallon can we use utterance_end_ms=1000&interim_results=true for streaming ? I saw that utterances are not available for the streaming.

Thanks!

**Reply by nikolawhallon** [reply]
The utterance end feature is different than the utterance feature - utterance_end_ms is a feature exclusive to streaming. Sorry for the confusing nomenclature - lots of overloaded terms..

**Reply by Fujiwaranosai** [reply]
@nikolawhallon I used the latest NodeJs SDK and configured the utterance_end_ms like below:

`const deepgramLive: LiveClient = this.deepgram.listen.live({
      language: 'en',
      smart_format: true,
      interim_results: true,
      model: 'nova',
      encoding: 'mulaw',
      sample_rate: 8000,
      endpointing: 300,
      utterance_end_ms: 1000,
    });`

but I didn't get any message with this format no matter how long I waited (when there was no transcript detected and the speech_final hasn't arrived in several seconds):

`{"type":"UtteranceEnd"}`

Do you have any idea?

**Reply by nikolawhallon** [reply]
I don't think the feature is supported by the NodeJs SDK, it only just recently got included in the Python and GO SDKs. It has been available via the API for quite some time. There is an active task to update the NodeJs SDK, but in the meantime this user made a fork which supports the feature:
https://github.com/diegoasua/deepgram-js-sdk

the diff is quite reasonable:
https://github.com/deepgram/deepgram-js-sdk/commit/55210c84e93fc75a291bcac7a153c134d974b842
